odoo/addons/base/models/ir_model.py

```diff
@@ -1491,6 +1492,35 @@ def group_names_with_access(self, model_name, access_mode):
         """ % access_mode, [model_name])
         return [('%s/%s' % x) if x[0] else x[1] for x in self._cr.fetchall()]
 
+    @tools.ormcache('self.env.uid')
+    def _acl(self):
+        # our base should be the model so we can get models for which there is no ir.model.acces
+        self._cr.execute("""
+            -- TODO: should only return gu.uid or !gu.uid, but not sure how
+            SELECT m.model, gu.uid,
+                bool_or(coalesce(a.perm_read, false)),
+                bool_or(coalesce(a.perm_write, false)),
+                bool_or(coalesce(a.perm_create, false)),
+                bool_or(coalesce(a.perm_unlink, false))
+            FROM ir_model m
+            JOIN ir_model_access a ON (a.model_id = m.id)
+            LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
+            WHERE a.active
+              AND (a.group_id IS NULL OR gu.uid = %s)
+            GROUP BY m.model, gu.uid
+        """, [self.env.uid])
+        # couldn't wrangle query into always returning a row per model so this
```

Comment:
to get results even for models w/o acl, how about a LEFT JOIN with a check on NULL id (also some suggested formatting)

```sql
   SELECT m.model, gu.uid,
          bool_or(coalesce(a.perm_read, false)) perm_read,
          bool_or(coalesce(a.perm_write, false)) perm_write,
          bool_or(coalesce(a.perm_create, false)) perm_create,
          bool_or(coalesce(a.perm_unlink, false)) perm_unlink
     FROM ir_model m
LEFT JOIN ir_model_access a ON (a.model_id = m.id)
LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
    WHERE (a.id IS NULL or a.active) AND 
          (a.group_id IS NULL OR gu.uid = %s)
 GROUP BY m.model, gu.uid
```

Reply:
Doesn't seem to work. On a db with 104 models, this query returns 71 unique models (78 rows total).

On a much larger db (1076 models) it returns 1140 rows, but only 1017 unique models.

Reply:
Right, it seemed too obvious ;-) I was testing with an internal user, but permissions for portal users had a lot of missing entries.


> It might work by joining to a subquery though, if we first get only the IMA linked to the current user then join that into the main query?

Should work indeed, e.g. with 2 CTEs, but that gives a convoluted query. Maybe there's also not much speed benefit vs the Python version?

```sql
WITH group_specific_acl AS (
       SELECT m.model,
              bool_or(coalesce(a.perm_read, false)) r,
              bool_or(coalesce(a.perm_write, false)) w,
              bool_or(coalesce(a.perm_create, false)) c,
              bool_or(coalesce(a.perm_unlink, false)) u
         FROM ir_model m
         JOIN ir_model_access a ON (a.model_id = m.id)
         JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
        WHERE a.active AND gu.uid = 4 
     GROUP BY m.model
),
default_acl AS (
       SELECT m.model,
              bool_or(coalesce(a.perm_read, false)) r,
              bool_or(coalesce(a.perm_write, false)) w,
              bool_or(coalesce(a.perm_create, false)) c,
              bool_or(coalesce(a.perm_unlink, false)) u
         FROM ir_model m
         JOIN ir_model_access a ON (a.model_id = m.id)
        WHERE a.group_id IS NULL
     GROUP BY m.model
)

   SELECT m.model,
          coalesce(spec.r, false) OR coalesce(def.r, false) r,
          coalesce(spec.w, false) OR coalesce(def.w, false) w,
          coalesce(spec.c, false) OR coalesce(def.c, false) c,
          coalesce(spec.u, false) OR coalesce(def.u, false) u
     FROM ir_model m
LEFT JOIN group_specific_acl spec USING (model)
LEFT JOIN default_acl def USING (model)
```

Reply:
> Should work indeed, e.g. with 2 CTEs

I think (not quire sure yet) it might also work with a 2-nested query e.g. 
```sql
    SELECT m.model, a.perm_read, a.perm_write
    FROM ir_model m
    LEFT JOIN (
        SELECT distinct on (acl.model_id) acl.*
        FROM (
            SELECT a.model_id as model_id, bool_or(a.perm_read) as perm_read, bool_or(a.perm_write) as perm_write
            FROM ir_model_access a
            LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
            WHERE (a.group_id IS NULL OR gu.uid = %s)
            GROUP BY a.model_id, gu.uid
            ORDER BY a.model_id, gu.uid NULLS LAST
        ) as acl
    ) as a ON (m.id = a.model_id)
```

Reply:
@odony added a commit with a simpler version, and simplified the baseline as well (in case we don't take the risk of stripping it out to just delegate to the batch one in stable).